src/java.base/share/classes/sun/security/provider/NamedKeyPairGenerator.java

```diff
@@ -0,0 +1,145 @@
+/// a local type, and this parsed key will be passed to an operational method
+/// (For example, `implSign`) later. An implementation must not retain
+/// a reference of the parsed key.
+public abstract class NamedKeyPairGenerator extends KeyPairGeneratorSpi {
+
+    private final String fname; // family name
+    private final String[] pnames; // allowed parameter set name (at least one)
+
+    protected String name = null; // init as
+    private SecureRandom secureRandom;
+
+    /// Creates a new `NamedKeyPairGenerator` object.
+    ///
+    /// @param fname the family name
+    /// @param pnames supported parameter set names, at least one is needed.
+    ///     If multiple, the first one becomes the default parameter set name.
+    protected NamedKeyPairGenerator(String fname, String... pnames) {
+        this.fname = Objects.requireNonNull(fname);
+        if (pnames == null || pnames.length == 0) {
+            throw new AssertionError("pnames cannot be null or empty");
+        }
+        this.pnames = pnames;
+    }
+
+    private String checkName(String name) throws InvalidAlgorithmParameterException  {
+        for (var pname : pnames) {
+            if (pname.equalsIgnoreCase(name)) {
+                // return the stored standard name
+                return pname;
+            }
+        }
+        throw new InvalidAlgorithmParameterException(
+                "Unknown parameter set name: " + name);
+    }
+
+    @Override
+    public void initialize(AlgorithmParameterSpec params, SecureRandom random)
+            throws InvalidAlgorithmParameterException {
+        if (params instanceof NamedParameterSpec spec) {
+            name = checkName(spec.getName());
+        } else {
+            throw new InvalidAlgorithmParameterException(
+                    "Unknown AlgorithmParameterSpec: " + params);
+        }
+        this.secureRandom = random ;
+    }
+
+    @Override
+    public void initialize(int keysize, SecureRandom random) {
+        if (keysize != -1) {
+            // Bonus: a chance to provide a SecureRandom without
+            // specifying a parameter set name
+            throw new InvalidParameterException("keysize not supported");
+        }
+        this.secureRandom = random;
+    }
+
+    @Override
+    public KeyPair generateKeyPair() {
+        String pname = name != null ? name : pnames[0];
+        var keys = implGenerateKeyPair(pname, secureRandom);
+        return new KeyPair(new NamedX509Key(fname, pname, keys[0]),
+                new NamedPKCS8Key(fname, pname, keys[1]));
+    }
+
+    /// User-defined key pair generator.
+    ///
+    /// @param pname parameter set name
+    /// @param sr `SecureRandom` object, `null` if not initialized
+    /// @return public key and private key (in this order) in raw bytes
+    /// @throws ProviderException if there is an internal error
+    public abstract byte[][] implGenerateKeyPair(String pname, SecureRandom sr);
```

Comment:
Should be protected?

Reply:
I left them public because I have a local test that uses those method. This is definitely not a legitimate justification. I'll make them protected.